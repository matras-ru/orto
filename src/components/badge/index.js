export { default as CBadge } from './badge';
