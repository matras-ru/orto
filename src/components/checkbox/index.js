export { default as CCheckbox } from './checkbox';
export { default as CCheckboxGroup } from './checkbox-group';
