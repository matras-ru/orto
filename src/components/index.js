import CButton from '@/components/CButton';
import CIcon from '@/components/CIcon';
import CInput from '@/components/CInput';
import CLabel from '@/components/CLabel';
import CCheckbox from '@/components/CCheckbox';
import CCheckboxGroup from '@/components/CCheckboxGroup';
import CRadio from '@/components/CRadio';
import CRadioGroup from '@/components/CRadioGroup';

export {
    CButton,
    CError,
    CForm,
    CFormGroup,
    CIcon,
    CInput,
    CLabel,
    CCheckbox,
    CCheckboxGroup,
    CRadio,
    CRadioGroup
};
