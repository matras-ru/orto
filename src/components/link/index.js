export { default as CLink } from './link';
