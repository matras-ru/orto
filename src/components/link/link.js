import { getComponentConfig } from '@/config';
import { commonAttributes, install } from '@/mixins';
import { getHashMapValue } from '@/utils';

import defaultTheme from '@/themes/default/CLink';

const NAME = 'CLink';
const ANCHOR_TAG = 'a';
const validVariants = ['primary'];

const concat = (...args) => Array.prototype.concat.apply([], args);
const isRouterLink = tag => tag.toString().toLowerCase() !== ANCHOR_TAG;

const computeTag = ({ to, disabled } = {}, instance) => {
    return instance.$router && to && !disabled
        ? instance.$nuxt
            ? 'nuxt-link'
            : 'router-link'
        : ANCHOR_TAG;
};

const computeRel = ({ target, rel } = {}) => {
    if (target === '_blank' && isNull(rel)) {
        return 'noopener';
    }
    return rel || null;
};

const createThemeMap = ({ defaultClass, primaryClass }) => {
    return {
        variants: {
            primary: primaryClass,
            default: defaultClass
        }
    };
};

export const createProps = () => {
    return {
        href: {
            type: String,
            default: null
        },

        target: {
            type: String,
            default: () => getComponentConfig(NAME, 'target')
        },

        rel: {
            type: String,
            default: null
        },

        // router specific props ref: https://router.vuejs.org/api/#router-link
        to: {
            type: [String, Object],
            default: null
        },

        append: {
            type: Boolean,
            default: false
        },

        event: {
            type: [String, Array],
            default: 'click'
        },

        replace: {
            type: Boolean,
            default: false
        },

        activeClass: {
            type: String,
            default: 'router-link-active'
        },

        routerTag: {
            type: String,
            default: 'a'
        },

        exact: {
            type: Boolean,
            default: false
        },

        exactActiveClass: {
            type: String,
            default: 'router-link-exact-active'
        },

        // nuxt-link specific prop(s)
        noPrefetch: {
            type: Boolean,
            default: false
        }
    };
};

export default {
    name: NAME,
    inheritAttrs: false,
    ...install,
    props: {
        ...commonAttributes.props,
        ...createProps(),

        theme: {
            type: Object,
            default: () => defaultTheme
        },

        variant: {
            type: String,
            default: () => getComponentConfig(NAME, 'variant'),
            validator: value => validVariants.includes(value)
        },

        border: {
            type: Boolean,
            default: false
        },

        label: {
            type: String,
            default: null
        }
    },

    computed: {
        computedTag() {
            return computeTag({ to: this.to, disabled: this.disabled }, this);
        },

        isRouterLink() {
            return isRouterLink(this.computedTag);
        },

        computedRel() {
            return computeRel({ target: this.target, rel: this.rel });
        },

        computedProps() {
            return this.isRouterLink ? { ...this.$props, tag: this.routerTag } : {};
        },

        computedClass() {
            const { baseClass, disabledClass } = this.theme;
            const classes = [baseClass];

            if (this.disabled) {
                classes.push(disabledClass);
            }

            const { variants } = createThemeMap(this.theme);
            classes.push(getHashMapValue(variants, this.variant));

            return classes;
        }
    },

    methods: {
        onClick(e) {
            const isEvent = e instanceof Event;
            const suppliedHandler = this.$listeners.click;

            if (isEvent && this.disabled) {
                e.stopPropagation();
                e.stopImmediatePropagation();
            } else {
                if (this.isRouterLink && e.currentTarget.__vue__) {
                    e.currentTarget.__vue__.$emit('click', e);
                }

                // Call the suppliedHandler(s), if any provided
                concat(suppliedHandler)
                    .filter(h => h instanceof Function)
                    .forEach(handler => {
                        handler(...arguments);
                    });

                // Emit the global $root click event
                this.$root.$emit('clicked::link', e);
            }

            // Stop scroll-to-top behavior or navigation on
            // regular links when href is just '#'
            if (isEvent && (this.disabled || (!this.isRouterLink && this.href === '#'))) {
                e.preventDefault();
            }
        }
    },

    render(h) {
        const componentData = {
            class: this.computedClass,
            props: this.computedProps,
            attrs: {
                ...this.$attrs,
                rel: this.computeRel,
                target: this.target,
                tabindex: this.disabled ? '-1' : this.$attrs.tabindex ? this.$attrs.tabindex : null,
                'aria-disabled': this.disabled ? 'true' : null
            },

            [this.isRouterLink ? 'nativeOn' : 'on']: {
                ...this.$listeners,
                click: this.onClick
            }
        };

        if (this.href) {
            componentData.attrs.href = this.href;
        } else {
            delete componentData.props.href;
        }

        return h(this.computedTag, componentData, this.label ? this.label : this.$slots.default);
    }
};
