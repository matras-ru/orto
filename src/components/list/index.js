export { default as CList } from './list';
export { default as CListItem } from './list-item';
