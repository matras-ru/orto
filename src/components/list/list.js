import { mergeData } from 'vue-functional-data-merge';
import { install } from '@/mixins';

import ThemeClass from '@/themes/default/CList';
import { getComponentConfig } from '@/config';

const { baseClass, defaultClass, horizontalClass, startClass, endClass, betweenClass } = ThemeClass; // TODO: flex-class вынести

const NAME = 'CList';

const props = {
    tag: {
        type: String,
        default: () => getComponentConfig(NAME, 'tag')
    },

    horizontal: {
        type: Boolean,
        default: () => getComponentConfig(NAME, 'horizontal')
    },

    justify: {
        type: String,
        default: null
    }
};

const currentClass = ({ horizontal, justify }) => {
    const classes = [baseClass];

    // horizontal/vertical
    classes.push(horizontal ? [horizontalClass] : [defaultClass]);

    // horizontal align
    switch (justify) {
        case 'end':
            classes.push(endClass);
            break;
        case 'between':
            classes.push(betweenClass);
            break;
        case 'start':
            classes.push(startClass);
            break;
        default:
            break;
    }

    return classes;
};

export default {
    name: NAME,

    functional: true,

    ...install,

    props,

    render(h, { props, data, children }) {
        const componentData = {
            class: currentClass(props)
        };

        return h(props.tag, mergeData(data, componentData), children);
    }
};
