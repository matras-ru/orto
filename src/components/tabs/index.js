export { default as CTabs } from './tabs';
export { default as CTab } from './tab';
export { default as CTabPanels } from './tab-panels';
export { default as CTabPanel } from './tab-panel';
