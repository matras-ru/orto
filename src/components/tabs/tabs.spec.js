/* eslint-env jest */
import Vue from 'vue';
import { mount, config } from '@vue/test-utils';

import CTabs from './tabs';
import CTab from './tab';

import { CList, CListItem } from '@/components/list';
import { CLink } from '@/components/link';

config.stubs['CList'] = CList;
config.stubs['CListItem'] = CListItem;
config.stubs['CLink'] = CLink;

const baseTabLinkClasses =
    'outline-none select-none inline-block no-underline cursor-pointer font-semibold text-lg uppercase px-1-5 py-0-7';
const activeTabLinkClasses = 'text-secondary-200 border-b-4 border-secondary-200';

describe('CTabs', () => {
    it('default', () => {
        const wrapper = mount(CTabs);

        expect(wrapper.element.tagName).toEqual('UL');
        expect(wrapper.attributes('role')).toBeDefined();
        expect(wrapper.attributes('role')).toBe('tablist');

        expect(wrapper.html()).toMatchSnapshot();

        wrapper.destroy();
    });

    it('custom props', () => {
        const wrapper = mount(CTabs, {
            context: {
                props: {
                    vertical: true,
                    justify: 'center'
                }
            }
        });

        expect(wrapper.classes().length).toBe(3);
        expect(wrapper.classes()).toContain('flex-col');
        expect(wrapper.classes()).toContain('justify-center');

        wrapper.destroy();
    });

    it('with tabs', () => {
        const wrapper = mount(CTabs, {
            slots: { default: [CTab, CTab, CTab] }
        });

        expect(wrapper).toBeDefined();
        expect(wrapper.findAll('li[role="presentation"]').length).toBe(3);

        wrapper.destroy();
    });

    it('tab link style: default, active', () => {
        const App = Vue.extend({
            data() {
                return {
                    tabsModel: 'tab-2'
                };
            },
            render(h) {
                return h(
                    CTabs,
                    {
                        props: {
                            modelValue: this.tabsModel
                        }
                    },
                    [
                        h(CTab, { props: { name: 'tab-1' }, ref: 'tab-1' }, 'tab 1'),
                        h(CTab, { props: { name: 'tab-2' }, ref: 'tab-2' }, 'tab 2')
                    ]
                );
            }
        });

        const wrapper = mount(App);

        const tab1 = wrapper.findComponent({ ref: 'tab-1' }); // default tab
        const tab2 = wrapper.findComponent({ ref: 'tab-2' }); // active tab

        expect(wrapper).toBeDefined();
        expect(wrapper.findAllComponents(CTab).length).toBe(2);

        expect(tab1.find('a').classes().sort()).toEqual(`${baseTabLinkClasses}`.split(' ').sort());

        expect(tab2.find('a').classes().sort()).toEqual(
            `${baseTabLinkClasses} ${activeTabLinkClasses}`.split(' ').sort()
        );

        expect(wrapper.html()).toMatchSnapshot();

        wrapper.destroy();
    });

    it('change tab', () => {
        const App = Vue.extend({
            data() {
                return {
                    tabsModel: 'tab-1'
                };
            },

            render(h) {
                return h(
                    CTabs,
                    {
                        props: {
                            modelValue: this.tabsModel // v-model
                        },
                        on: {
                            onChange: val => (this.tabsModel = val) // v-model
                        }
                    },
                    [
                        h(CTab, { props: { name: 'tab-1' }, ref: 'tab-1' }, 'tab 1'),
                        h(CTab, { props: { name: 'tab-2' }, ref: 'tab-2' }, 'tab 2')
                    ]
                );
            }
        });

        const wrapper = mount(App);
        const tab2 = wrapper.findComponent({ ref: 'tab-2' }); // active tab

        expect(wrapper).toBeDefined();

        expect(wrapper.vm.tabsModel).toBe('tab-1');
        tab2.find('a').trigger('click');
        expect(wrapper.vm.tabsModel).toBe('tab-2');

        wrapper.destroy();
    });
});
