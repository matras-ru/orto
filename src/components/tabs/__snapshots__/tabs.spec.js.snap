// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CTabs default 1`] = `<ul role="tablist" class="flex flex-row justify-start"></ul>`;

exports[`CTabs tab link style: default, active 1`] = `
<ul role="tablist" class="flex flex-row justify-start">
  <li role="presentation" class="inline-block"><a role="tab" aria-selected="false" tabindex="-1" id="tab-tab-1" href="#tab-1" target="_self" class="outline-none select-none font-semibold text-lg uppercase px-1-5 py-0-7 no-underline cursor-pointer inline-block">tab 1</a></li>
  <li role="presentation" class="inline-block"><a role="tab" aria-selected="true" id="tab-tab-2" href="#tab-2" target="_self" class="outline-none select-none font-semibold text-lg uppercase px-1-5 py-0-7 no-underline cursor-pointer inline-block text-secondary-200 border-b-4 border-secondary-200">tab 2</a></li>
</ul>
`;
